src/engine/objects/robot/sensors/SimDistanceSensor.ts

```diff
@@ -119,24 +126,75 @@ export class SimDistanceSensor extends SimBasicSensor {
     return this._detectAngle;
   }
 
+  /**
+   * Sensor output
+   *
+   * This value represents the percentage of V<sub>ref</sub> corresponding
+   * to the distance detected
+   */
+  get value(): number {
+    const pct =
+      (this.getDistance() - this.minRange) / (this.maxRange - this.minRange);
+    return pct;
+  }
+
+  private getDistance(): number {
+    let result = 0;
+
+    const sensorAngle = -this._body.getAngle();
+    const bodyCenter = this._body.getWorldCenter();
+
+    // Set up min/max points
+    const p1 = new Vec2(
+      bodyCenter.x + Math.sin(sensorAngle) * this.minRange,
+      bodyCenter.y + Math.cos(sensorAngle) * this.minRange
+    );
+
+    // TODO Implement cone of detection here
+    const p2 = new Vec2(
+      bodyCenter.x + Math.sin(sensorAngle) * this.maxRange,
+      bodyCenter.y + Math.cos(sensorAngle) * this.maxRange
+    );
+
+    const detectionRange = this.maxRange - this.minRange;
+
+    this._body
+      .getWorld()
+      .rayCast(
+        p1,
+        p2,
+        (fixture: Fixture, p: Vec2, normal: Vec2, fraction: number): number => {
+          // Ignore everything that is part of the robot that this sensor is attached to
+          if (fixture.getUserData()) {
+            const userData = fixture.getUserData() as ISimUserData;
+            if (userData.robotGuid === this._robotGuid) {
+              return -1;
+            }
+          }
+
+          // Set the result to be the current reading
+          result = fraction * detectionRange;
+
+          return fraction;
+        }
+      );
+
+    return result;
```

Comment:
Does this have the effect of making no object found return distance of 0? i think there shoud be some max value we can return or maybe infinity.? thoughts?

Reply:
Really depends on the kind of sensor. In the demo I made sure to check that the value was above 0. For real life sensors, it'll depend on how they are implemented. Commonly used IR sensor like this: https://www.pololu.com/file/0J85/gp2y0a21yk0f.pdf (see page 5) have a range from 0 to 3.5V, with objects further than detection range returning something along the lines of 0.5V.

Generally, if there's no object detected, a sensor value could go either way, so it'll be up to the robot code to check for appropriate values.

I do want to add in an optional conversion function that will take the measured distance and tweak the return values to match the physical characteristics of a real sensor. E.g. take the reported distance and convert it against the graph in the datasheet